File: 백준/Gold/1174. 줄어드는 수/줄어드는 수.java
import java.io.*;
import java.util.*;

public class Main {
    static ArrayList<Long> nums = new ArrayList<>();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());

        for (int i = 0; i < 10; i++) {
            backtrack(i, i);
        }

        if (N > nums.size()) {
            System.out.println(-1);
        } else {
            nums.sort(Comparator.naturalOrder());
            System.out.println(nums.get(N-1));
        }
    }

    static void backtrack(long n, int last) {
        nums.add(n);

        for (int i = 0; i < last; i++) {
            backtrack(n*10 + i, i);
        }
    }
}

File: 백준/Gold/1174. 줄어드는 수/줄어드는 수.py
# 줄어드는 수
import sys
input = sys.stdin.readline

def backtrack(n, last):
    nums.append(n)

    for i in range(0, last):
        backtrack(n*10+i, i)

N = int(input())
nums = []

for i in range(0, 10):
    backtrack(i, i)

if N > len(nums):
    print(-1)
else:
    nums.sort()
    print(nums[N-1])
